File: scripts/group_cluster.py
import heapq

def get_max_cluster_size(booting_power, processing_power, power_max):
    n = len(booting_power)
    pre_sum = [0] * (n + 1)
    for i, e in enumerate(processing_power):
        pre_sum[i + 1] = pre_sum[i] + e

    print(pre_sum)
    max_heap = []
    max_size = 0
    start = 0
    end = 0
    while start <= end < n:
        heapq.heappush(max_heap, -booting_power[end])
        cur_power = - max_heap[0] + (pre_sum[end + 1] - pre_sum[start]) * (end - start + 1)
        if cur_power <= power_max:
            max_size = max(max_size, (end - start + 1))
            end += 1
        else:
            max_heap.remove(-booting_power[start])
            heapq.heapify(max_heap)
            start += 1
    return max_size


if __name__ == '__main__':
    res = get_max_cluster_size([3, 6, 1, 3, 4], [2, 1, 3, 4, 5], 25)
    print(res)

File: src/com/xinxin/algo/basic/amazon/GroupCluster.java
package com.xinxin.algo.basic.amazon;

import java.util.PriorityQueue;

public class GroupCluster {
    public static void main(String[] args) {
        int res = maxSustainableCluster(new int[] {3,6,1,3,4}, new int[] {2,1,3,4,5}, 25);
        System.out.println(res);
    }

    // 3,6,1,3,4
    public static int maxSustainableCluster(int[]  bootingPower, int[] processingPower, int powerMax) {
        int n = bootingPower.length;
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a,b) -> b - a);
        int[] preSum = new int[n+1];
        for (int i = 1; i < n; ++ i) {
            preSum[i] = preSum[i - 1] + processingPower[i];
        }

        int start = 0;
        int end = 0;
        int maxGroupSize = 0;

        while (start <= end && end < n) {
            maxHeap.add(bootingPower[end]);
            int maxBootingPower = maxHeap.peek();
            int curPower = maxBootingPower + (preSum[end + 1] - preSum[start]) * (end - start + 1);
            if (curPower <= powerMax) {
                maxGroupSize = Math.max(maxGroupSize, end - start + 1);
                ++ end;
            } else {
                maxHeap.remove(bootingPower[start]);
                ++ start;
            }
        }
        return maxGroupSize;
    }
}
